pkg/ndctl/region.go

```diff
@@ -208,13 +204,36 @@ func (r *region) CreateNamespace(opts CreateNamespaceOpts) (Namespace, error) {
 	}
 
 	if opts.Size != 0 {
-		ways := uint64(C.ndctl_region_get_interleave_ways(r))
-		align = align * ways
-		if opts.Size%align != 0 {
-			// Round up size to align with next block boundary.
-			opts.Size = (opts.Size/align + 1) * align
-			klog.V(4).Infof("%s: namespace size must align to interleave-width:%d * alignment:%d, force-align to %d",
-				regionName, ways, align, opts.Size)
+		/* old method, using namespace align and interleave ways.
+		If we use fixed 2M namespace align, this is always overshadowed
+		by region align which is 16 or 96 MB so we may leave this code out
+		ways := r.InterleaveWays()
```

Comment:
I'd prefer to not make that assumption. Better align with both to be on the safe side.

Reply:
More specifically, check the alignment requirement of the pfn (= `ndctl_pfn_get_align`). If that is not set, then we can fall back to some builtin default, like this code does:
https://github.com/pmem/ndctl/blob/ea014c0c9ec8d0ef945d072dcc52b306c7a686f9/ndctl/namespace.c#L799-L814

Reply:
ndctl also checks both:
https://github.com/pmem/ndctl/blob/ea014c0c9ec8d0ef945d072dcc52b306c7a686f9/ndctl/namespace.c#L838-L843

Reply:
I was thinking about that, too. Its simple to find common align value if one is 2G, but what if some day was need to switch to 1G, then what is common align between 1G and 96MB, isnt it getting quite large?

Reply:
hm I meant, if one is 2MB

Reply:
Although there it aborts when not aligned. We need to decrease the size.

Reply:
Wait, *increase* the size.

Reply:
In the GetCapacity call we round down. When creating, we round up.

Reply:
If we have regionAlign and namespaceAlign, then we need to align at the smallest common multiple of both. Then the resulting size will be a multiple of both.

Reply:
I most cases, the smallest common multiple will be the region alignment. We just don't make that assumption by actually doing the math.

Reply:
> In the GetCapacity call we round down. When creating, we round up.

in current draft, in creation we align down. idea is, if we give up compensating idea, it does not matter much, and it is safer to align down, as aligning up has risk to exceed total size, or percentage what is allowed. Total sizes seem not to be aligned, which is bit unfriendly in that respect

Reply:
> but what if some day was need to switch to 1G, then what is common align between 1G and 96MB, isnt it getting quite large

And? If the requirement is to align with both, we have no choice.

> in current draft, in creation we align down. idea is, if we give up compensating idea, it does not matter much, 

It does matter. Our promise to the user is that when they request a volume of size 1GB, we use that amount of PMEM, not less.

> aligning up has risk to exceed total size

In that case the host does not have enough capacity left for the volume and it needs to be created elsewhere. Rounding down in GetCapacity is supposed to avoid this scenario.

Reply:
> hm I meant, if one is 2MB

The smallest common multiple of 2Mi and 1Gi is 1Gi.

Reply:
I meant, if in theory, we need to switch to 1GB namespace align again, then for example LCM of 96 MB and 1GB will be 3072 MB. But does not matter, I added LCM code that calculates it in all cases. I hope borrowing LCM function from Go playground is allowed, with added comment of course.

Reply:
> I meant, if in theory, we need to switch to 1GB namespace align again, then for example LCM of 96 MB and 1GB will be 3072 MB.

If someone configures their hardware so that these two alignment constraints apply, then yes, that's what we'll have to do.